### ISSUE TITLE
vdo-larod + opencv

### ISSUE THREAD

**Issue by Shaiken** [AxisCommunications/acap-native-sdk-examples#256]
Is there a complete Dockerfile + app/example.cpp example of opencv + vdo-larod?

**Comment by pataxis** [comment]
Hi @Shaiken , no there's not. There's a close one though in the [using-opencv](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/using-opencv) example that combines OpenCV and VDO.

**Comment by Shaiken** [comment]
I'm new to this, can you help me check what I merged?
I tried using openai and copilot, but still ran into some problems.
I just want to try the vdo-larod example and just draw a box on the image.

Makefile
`PROG1  = vdo_larod
OBJS1  = $(PROG1).cpp utility-functions.c $(wildcard *.cpp)
PROGS  = $(PROG1)

PKGS = gio-2.0 vdostream gio-unix-2.0 liblarod

CFLAGS += -Os -pipe -std=c++11
CFLAGS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) pkg-config --cflags $(PKGS))
LDLIBS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) pkg-config --libs $(PKGS))
LDLIBS += -lm -lopencv_video -lopencv_imgproc -lopencv_core -lpthread

CFLAGS += -I$(SDKTARGETSYSROOT)/usr/include/opencv4
CFLAGS += -Wall
# This example is based on larod version 3
CFLAGS += -DLAROD_API_VERSION_3

LDFLAGS = -L./lib -Wl,--no-as-needed,-rpath,'$$ORIGIN/lib'

all: $(PROGS)

$(PROG1): $(OBJS1)
	$(CC) $(CXX) $(CFLAGS) $(LDFLAGS) $^ $(LDLIBS) -o $@ ; \
	$(STRIP) --strip-unneeded $@

clean:
	rm -f $(PROGS) *.o *.eap`


Dockerfile
`ARG ARCH=aarch64
ARG VERSION=1.13
ARG UBUNTU_VERSION=22.04
ARG REPO=axisecp
ARG SDK=acap-native-sdk

FROM ${REPO}/${SDK}:${VERSION}-${ARCH}-ubuntu${UBUNTU_VERSION}

RUN apt-get update && apt-get install -y --no-install-recommends \
    unzip \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

WORKDIR /opt/app/
ARG CHIP

# Download the pretrained model
ARG MODEL_BUCKET=https://acap-artifacts.s3.eu-north-1.amazonaws.com/models
RUN if [ "$CHIP" = artpec8 ]; then \
        curl -o model.zip $MODEL_BUCKET/models.aarch64.artpec8.zip ; \
    elif [ "$CHIP" = cpu ]; then \
        curl -o model.zip $MODEL_BUCKET/models.aarch64.artpec8.zip ; \
    elif [ "$CHIP" = edgetpu ]; then \
        curl -o model.zip $MODEL_BUCKET/models.armv7hf.edgetpu.zip ; \
    elif [ "$CHIP" = cv25 ]; then \
        curl -o model.zip $MODEL_BUCKET/tensorflow_to_larod_models.cv25.zip ; \
    else \
        printf "Error: '%s' is not a valid value for the CHIP variable\n", "$CHIP"; \
        exit 1; \
    fi && \
    unzip -q model.zip && rm -f model.zip

# Copy the library to application folder
WORKDIR /opt/app
COPY ./app .

# Set general arguments
ARG ARCH
ARG SDK_LIB_PATH_BASE=/opt/axis/acapsdk/sysroots/${ARCH}/usr
ARG BUILD_DIR=/opt/build

#-------------------------------------------------------------------------------
# Prepare build environment
#-------------------------------------------------------------------------------

# Install build dependencies for cross compiling OpenCV
RUN DEBIAN_FRONTEND=noninteractive \
    apt-get update && apt-get install -y -f --no-install-recommends \
    cmake && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

#-------------------------------------------------------------------------------
# Build OpenCV libraries
#-------------------------------------------------------------------------------

ARG OPENCV_VERSION=4.5.1
ARG OPENCV_DIR=${BUILD_DIR}/opencv
ARG OPENCV_SRC_DIR=${OPENCV_DIR}/opencv-${OPENCV_VERSION}
ARG OPENCV_BUILD_DIR=${OPENCV_DIR}/build

WORKDIR ${OPENCV_DIR}
SHELL ["/bin/bash", "-o", "pipefail", "-c"]
RUN curl -fsSL https://github.com/opencv/opencv/archive/$OPENCV_VERSION.tar.gz | tar -xz

WORKDIR ${OPENCV_BUILD_DIR}
ENV COMMON_CMAKE_FLAGS="-S $OPENCV_SRC_DIR \
        -B $OPENCV_BUILD_DIR \
        -D CMAKE_INSTALL_PREFIX=$SDK_LIB_PATH_BASE \
        -D CMAKE_BUILD_TYPE=RELEASE \
        -D WITH_OPENEXR=OFF \
        -D WITH_GTK=OFF \
        -D WITH_V4L=OFF \
        -D WITH_FFMPEG=OFF \
        -D WITH_GSTREAMER=OFF \
        -D WITH_GSTREAMER_0_10=OFF \
        -D BUILD_LIST=core,imgproc,video \
        -D BUILD_EXAMPLES=OFF \
        -D BUILD_OPENCV_APPS=OFF \
        -D BUILD_DOCS=OFF \
        -D BUILD_JPEG=ON \
        -D BUILD_PNG=OFF \
        -D WITH_JASPER=OFF \
        -D BUILD_PROTOBUF=OFF \
        -D OPENCV_GENERATE_PKGCONFIG=ON "

# hadolint ignore=SC2086
RUN if [ "$ARCH" = armv7hf ]; then \
        # Source SDK environment to get cross compilation tools
        . /opt/axis/acapsdk/environment-setup* && \
        # Configure build with CMake
        cmake \
        -D CMAKE_CXX_COMPILER=${CXX%-g++*}-g++ \
        -D CMAKE_CXX_FLAGS="${CXX#*-g++}" \
        -D CMAKE_C_COMPILER=${CC%-gcc*}-gcc \
        -D CMAKE_C_FLAGS="${CC#*-gcc}" \
        -D CMAKE_TOOLCHAIN_FILE=${OPENCV_SRC_DIR}/platforms/linux/arm-gnueabi.toolchain.cmake \
        -D CPU_BASELINE=NEON,VFPV3 \
        -D ENABLE_NEON=ON \
        -D ENABLE_VFPV3=ON \
        $COMMON_CMAKE_FLAGS && \
        # Build and install OpenCV
        make -j "$(nproc)" install ; \
    elif [ "$ARCH" = aarch64 ]; then \
        # Source SDK environment to get cross compilation tools
        . /opt/axis/acapsdk/environment-setup* && \
        # Configure build with CMake
        # No need to set NEON and VFP for aarch64 since they are implicitly
        # present in an any standard armv8-a implementation.
        cmake \
        -D CMAKE_CXX_COMPILER=${CXX%-g++*}-g++ \
        -D CMAKE_CXX_FLAGS="${CXX#*-g++}" \
        -D CMAKE_C_COMPILER=${CC%-gcc*}-gcc \
        -D CMAKE_C_FLAGS="${CC#*-gcc}" \
        -D CMAKE_TOOLCHAIN_FILE=${OPENCV_SRC_DIR}/platforms/linux/aarch64-gnu.toolchain.cmake \
        $COMMON_CMAKE_FLAGS && \
        # Build and install OpenCV
        make -j "$(nproc)" install ; \
    else \
        printf "Error: '%s' is not a valid value for the ARCH variable\n", "$ARCH"; \
        exit 1; \
    fi

#-------------------------------------------------------------------------------
# Copy the built library files to application directory
#-------------------------------------------------------------------------------

WORKDIR /opt/app
COPY ./app .
RUN mkdir lib && \
    cp -P ${OPENCV_BUILD_DIR}/lib/lib*.so* ./lib/
    
#-------------------------------------------------------------------------------
# Finally build the ACAP application
#-------------------------------------------------------------------------------
    
#RUN . /opt/axis/acapsdk/environment-setup* && acap-build .
# Build the ACAP application
RUN cp /opt/app/manifest.json.${CHIP} /opt/app/manifest.json && \
    . /opt/axis/acapsdk/environment-setup* && \
    if [ "$CHIP" = artpec8 ] || [ "$CHIP" = cpu ]; then \
        acap-build . -a 'models/converted_model.tflite'; \
    elif [ "$CHIP" = edgetpu ]; then \
        acap-build . -a 'models/converted_model_edgetpu.tflite'; \
    elif [ "$CHIP" = cv25 ]; then \
        acap-build . -a 'models/car_human_model_cavalry.bin'; \
    else \
        printf "Error: '%s' is not a valid value for the CHIP variable\n", "$CHIP"; \
        exit 1; \
fi

`

**Comment by Shaiken** [comment]
I tried changing vdo-larod from c to cpp. When docker builds, if the parameters are not at the top, an error will be reported.   why...?

**Comment by Cacsjep** [comment]
Share the dockerfile

**Comment by Shaiken** [comment]
[Dockerfile.zip](https://github.com/user-attachments/files/15879568/Dockerfile.zip)

**Comment by Shaiken** [comment]
I have another question, how can I use opencv to display the drawn image in a Stream?

**Comment by Shaiken** [comment]
Thank you for your assistance. I've tried the examples from the acap-native-sdk-examples, and they all work. However, integrating them into my own Docker environment is a bit challenging. In addition to the Dockerfile, I also need to pay attention to the Makefile. Converting C to C++ and using VOD + OpenCV has not been very smooth because I want to use a piecewise model and cannot simply use axoverlay.

Later, I saw that acap-computer-vision-sdk-examples can use Flask for communication. Although it seems that this approach cannot match the speed of acap-native, I am unsure how to start implementing it.

**Comment by Cacsjep** [comment]
Whats your goal?

**Comment by Shaiken** [comment]
Thank you for your help. I want to run the tensorflow lite model, and it is a semantic segmentation model. I want to draw its boundaries and show it in the Installation.
![image](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/30071132/2c5e2921-a3ca-41c2-b6e1-b6a3f7bedd93)

**Comment by Cacsjep** [comment]
Lets Talk at discord https://discord.com/invite/EVwKjX8B

**Comment by Shaiken** [comment]
Can axoverlay do it? 
Because when I look at its examples, irregular borders are not drawn.
I went in

**Comment by Cacsjep** [comment]
could be done via axoverlay so this could be closed ;)

**Comment by pataxis** [comment]
There's also a new API for drawing [bounding boxes](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/bounding-box) on more chipsets, Axoverlay is for ARTPEC devices only.
